processing/src/main/java/org/apache/druid/segment/virtual/ExpressionSelectors.java

```diff
@@ -230,8 +230,9 @@ public static DimensionSelector makeDimensionSelector(
           && capabilities.getType() == ValueType.STRING
           && capabilities.isDictionaryEncoded()
           && capabilities.isComplete()
-          && !capabilities.hasMultipleValues()
-          && !exprDetails.getArrayColumns().contains(column)
+          && !exprDetails.hasInputArrays()
+          && !exprDetails.isOutputArray()
+          && (!capabilities.hasMultipleValues() || exprDetails.getFreeVariables().size() == 1)
       ) {
         // Optimization for dimension selectors that wrap a single underlying string column.
```

Comment:
The flag chain above is getting complicated!

Could you please alter this comment to reflect what it's looking for? If I understand it right, that should be something like:

> Optimization for dimension selectors that wrap a single underlying string column. The string column can be multi-valued, but if so, it must be implicitly mappable (i.e. the expression is not treating it as an array, not wanting to output an array, and the multi-value dimension appears exactly once).

By the way, I have to assume logic like this is being used elsewhere to implicitly map expressions that go through the non-optimized selector. Is there any opportunity to extract the logic into a common helper method? It would make it easier for readers and probably less error prone too.

Reply:
The logic is only used in these 2 places in ExpressionSelectors I think, and it's slightly different between the uses because `SingleStringInputDimensionSelector` supports multi-value inputs, but `SingleStringInputCachingExpressionColumnValueSelector` does not. I'll play around and see what I can do, or at least comment what's going on heavily.